File: 0509-fibonacci-number/0509-fibonacci-number.java
class Solution {
    private int ans(int n,Map<Integer,Integer> cache){
        if(n<2){
            return n;
        }
        
        int currKey = n;
        
        if(cache.containsKey(currKey)){
            return cache.get(currKey);
        }
        
        int firstHalf = ans(n-1,cache);
        int secondHalf = ans(n-2,cache);
        
        cache.put(currKey,firstHalf+secondHalf);
        
        return cache.get(currKey);
    }
    
    public int fib(int n) {
       return ans(n,new HashMap<Integer,Integer>());
    }
}

File: 0509-fibonacci-number/0509-fibonacci-number.py
class Solution:
    def calculateFibonaacci(self, n :int) -> int:
        if(n==0):
            return 0
        elif(n==1 or n==2):
            return 1
        
        return self.calculateFibonaacci(n-1) + self.calculateFibonaacci(n-2)
    
    def fib(self, n: int) -> int:
        if(n==0):
            return 0
        elif(n==1 or n==2):
            return 1
        return self.calculateFibonaacci(n-1) + self.calculateFibonaacci(n-2)
        